File: src/main/java/me/zero/jarpwner/plugin/AllatoriExpiryTransformer.java
package me.zero.jarpwner.plugin;

import me.zero.jarpwner.transform.Transformer;
import me.zero.jarpwner.util.Pattern;
import me.zero.jarpwner.util.Util;
import org.objectweb.asm.tree.AbstractInsnNode;
import org.objectweb.asm.tree.ClassNode;
import org.objectweb.asm.tree.TypeInsnNode;

import java.util.*;

import static org.objectweb.asm.Opcodes.*;

/**
 * @author Brady
 * @since 3/31/2019
 */
public class AllatoriExpiryTransformer implements Transformer {

    private static final Pattern PATTERN = Pattern.of(
            NEW,
            DUP,
            LDC,
            INVOKESPECIAL,
            NEW,
            DUP,
            INVOKESPECIAL,
            SWAP,
            INVOKEVIRTUAL,
            IFEQ,
            NEW,
            DUP,
            LDC,
            INVOKESTATIC,
            INVOKESPECIAL,
            ATHROW
    );

    private static final Pattern PATTERN_NO_OBF = Pattern.of(
            NEW,
            DUP,
            LDC,
            INVOKESPECIAL,
            NEW,
            DUP,
            INVOKESPECIAL,
            SWAP,
            INVOKEVIRTUAL,
            IFEQ,
            NEW,
            DUP,
            LDC,
            INVOKESPECIAL,
            ATHROW
    );

    private static final List<Pattern> PATTERNS = Arrays.asList(PATTERN, PATTERN_NO_OBF);

    private int removedMatches;

    @Override
    public final void apply(ClassNode cn) {
        cn.methods.forEach(mn -> PATTERNS.forEach(pattern -> pattern.find(mn.instructions, Pattern.SearchFlags.IGNORE_ALL).forEach(range -> {
            var insns = range.getAll(true);
            if (insns == null) {
                return;
            }

            var newInsns = insns.<TypeInsnNode>allWithOpcode(NEW);

            if (!"java/util/Date".equals(newInsns.get(0).desc))
                return;

            if (!"java/util/Date".equals(newInsns.get(1).desc))
                return;

            if (!"java/lang/Throwable".equals(newInsns.get(2).desc))
                return;

            if (!"java/util/Date.after(Ljava/util/Date;)Z".equals(Util.getFullDesc(insns.firstWithOpcode(INVOKEVIRTUAL))))
                return;

            range.delete(mn.instructions);

            removedMatches++;
        })));
    }

    @Override
    public final boolean accepts(String path) {
        return true;
    }

    @Override
    public final Collection<String> getInfo() {
        return Collections.singletonList(
                String.format("Removed %d expiry check(s)", this.removedMatches)
        );
    }
}

File: src/main/kotlin/me/zero/jarpwner/plugin/AllatoriExpiryTransformer.kt
package me.zero.jarpwner.plugin

import me.zero.jarpwner.transform.Transformer
import me.zero.jarpwner.extension.completeDesc
import me.zero.jarpwner.extension.findPattern
import me.zero.jarpwner.extension.to
import org.objectweb.asm.Opcodes.*
import org.objectweb.asm.tree.*
import java.util.*

class AllatoriExpiryTransformer : Transformer {

    companion object {
        private val PATTERN: Array<Int?> = arrayOf(
            NEW,
            DUP,
            LDC,
            INVOKESPECIAL,
            NEW,
            DUP,
            -1,
            INVOKESPECIAL,
            SWAP,
            INVOKEVIRTUAL,
            IFEQ,
            NEW,
            DUP,
            LDC,
            INVOKESTATIC,
            INVOKESPECIAL,
            ATHROW
        )

        private val PATTERN_NO_OBF: Array<Int?> = arrayOf(
            NEW,
            DUP,
            LDC,
            INVOKESPECIAL,
            NEW,
            -1,
            DUP,
            INVOKESPECIAL,
            SWAP,
            INVOKEVIRTUAL,
            IFEQ,
            NEW,
            DUP,
            LDC,
            INVOKESPECIAL,
            ATHROW
        )
    }

    private var removedMatches = 0

    override fun apply(cn: ClassNode) {
        cn.methods.forEach { mn ->
            arrayOf(PATTERN, PATTERN_NO_OBF).forEach { pattern ->
                mn.instructions.findPattern(*pattern).forEach { range ->
                    val from  = range.first
                    val to    = range.second
                    val nodes = from.to(to, inclusive = true)

                    val dateExpiry  = nodes[0]  as TypeInsnNode
                    val dateCurrent = nodes[4]  as TypeInsnNode
                    val dateAfter   = nodes[9]  as MethodInsnNode
                    val throwable   = nodes[11] as TypeInsnNode

                    if (dateExpiry.desc          != "java/util/Date" ||
                        dateCurrent.desc         != "java/util/Date" ||
                        dateAfter.completeDesc() != "java/util/Date.after(Ljava/util/Date;)Z" ||
                        throwable.desc           != "java/lang/Throwable")
                        return

                    nodes.forEach(mn.instructions::remove)

                    removedMatches++
                }
            }
        }
    }

    override fun accepts(path: String): Boolean {
        return true
    }

    override fun getInfo() : Collection<String> {
        return Collections.singletonList("Removed ${this.removedMatches} expiry checks")
    }
}